### DISCUSSION TITLE
How to get the optimal basic solution of an LP?

### DISCUSSION THREAD

**Discussion by erelsgl** [cvxpy/cvxpy#1524] [Q&A] [answered]
I have a linear program, and I need to find its optimal BFS (=Basic Feasible Solution) - a solution that contains few nonzero elements (specifically: in an LP with m constraints, the BFS has at most m nonzero elements).

I saw that there is a "bfs" argument to MOSEK, but this is a commercial solver. Is there a similar option in other solvers?

I think it can be useful to add this parameter as a general argument to `solve`, since it can be useful regardless of the solver used.

I know that the Simplex method always returns a BFS. Is there a way to tell `solve` to use Simplex?

**Comment by SteveDiamond** [comment]
My understanding is that BFS is specific to linear programs. None of the default solvers (which are designed for quadratic programs or general cone programs) can provide a BFS.

**Reply by rileyjmurray** [reply]
@SteveDiamond with the exception of the new SciPy interface! It comes with two simplex-based algorithms for LP :)

**Comment by rileyjmurray** [comment, marked as answer]
You can select a simplex-based solver through our SciPy/HiGHS interface. Call
```
prob.solve(solver='SCIPY', scipy_options={'method': 'revised simplex'})
```
or 
```
prob.solve(solver='SCIPY', scipy_options={'method': 'highs-ds'})
```
See https://docs.scipy.org/doc/scipy/reference/optimize.linprog-highs-ds.html#optimize-linprog-highs-ds as a starting point if you want to look into SciPy methods / HiGHS more.

**Comment by erelsgl** [comment]
Thanks! When writing an application that needs a BFS, it may be more readable to send this as a parameter to solve, e.g.:

     prob.solve(return_bfs=True)

Then, `solve` could call the scipy solver with revised simplex (unless a different solver is specified).

**Reply by rileyjmurray** [reply]
@erelsgl that's a good point. The MOSEK interface already accepts a ``bfs`` keyword argument (``prob.solve(solver='MOSEK', bfs=True)``). So it would be reasonable for us to extend the semantics of that keyword. We'd need to make sure cvxpy has reasonable behavior when some does things like the following:
```
prob.solve(solver='GUROBI', bfs=True)
```
Which would ideally return a BFS. The GUROBI interface would need to be modified to support that.
Or a user might do something like
```
prob.solve(solver='SCS', bfs=True)
```
Which should raise a warning since SCS doesn't support basic feasible solutions.

There's also the question of if we'd want to create an order in which a solver is automatically selected if someone specifies ``bfs=True`` (i.e., maybe call a commercial solver if one is available before calling HiGHS).
